stats/bridge_stats.go

```diff
@@ -302,29 +297,40 @@ func (s *Service) readLoop(conn *connWrapper) {
 		// If the network packet is a system ping, respond to it directly
 		var ping string
 		if err := json.Unmarshal(blob, &ping); err == nil && strings.HasPrefix(ping, "primus::ping::") {
+			log.Info("[Bridge stats] The client receives ping from peers, should pong again.")
 			if err := conn.WriteJSON(strings.Replace(ping, "ping", "pong", -1)); err != nil {
 				log.Warn("Failed to respond to system ping message", "err", err)
 				return
 			}
 			continue
 		}
+
+		// Not a system ping, try to decode an actual state message
+		var msg map[string][]interface{}
+		if err := json.Unmarshal(blob, &msg); err != nil {
+			log.Warn("Failed to decode stats server message", "err", err)
+			return
+		}
+		log.Info("Received message from stats server", "msg", msg)
 	}
 }
 
-func (s *Service) setLastError(listener, err string) {
+func (s *Service) setLastError(listener, err string) error {
 	s.lock.Lock()
```

Comment:
This lock seems unnecessary, we can remove this in the struct. This lock is used by setLastError, setProcessedBlock and report before. But as these 3 functions are in different case of the same select, they cannot run concurrently so no synchronization is needed.

Reply:
Oke let me remove this lock

Reply:
Please remove this lock in the Service struct too